docs/design.md

```diff
@@ -0,0 +1,118 @@
+abstract state machine execution. This document covers the implementation
+of concord-bft. When the implementation differs from the paper, it will be
+mentioned explicitly.
+
+It is important to note that this document covers the code as it exists at the
+time of writing. When the code changes substantially this document should be
+updated.
+
+# Public Interfaces
+
+Concord-bft is a library meant to be used by application code. The external user
+API for the core engine is described by a [set of structs and abstract
+interfaces](../bftengine/include/bftengine). The file names are pretty self
+explanatory. Important structures will be mentioned explicitly below.
+
+All user interfaces live in the `bftEngine` namespace.
+
+# Clients
+A default client,
+[SimpleClient](../bftengine/include/bftengine/SimpleClient.hpp), is included
+with concord-bft. It is just one manner of implementing a client, and is
+implemented in C++ utilizing the `ICommunication` and `IReceiver` interfaces
+below. Additional clients can be implemented by utilizing the client messages
+described in the next subsection. In fact, we have a [client in
+Python](../util/pyclient/bft_client.py) that we use for testing that does just
+that.
+
+## Client Messages
+Client request and reply messages are prefixed with a `ClientRequestMsgHeader` or
+`ClientReplyMsgHeader` respectively. Each header, like all other messages in
+concord-bft is implemented as a packed C struct, meaning that there are no padding
+bytes between struct fields. This is accomplished via enclosing the structure
+definitions between `#pragma pack(push,1)` and `#pragma pack(pop)` lines.
+Packing ensures that the size of the structure is the sum of the size of the
+fields, and allows them to be sent over the wire and interpreted
+deterministically, given both machines have the same endianess.
+
+## Communication
+
+The `ICommunication` interface is used by the bftengine code to send
+asynchronous messages across the network. It is implemented for different
+transport layers such as UDP, TCP, and TLS in
+[bftengine/src/communication](../bftengine/src/communication).
+
+The `setReceiver` method is called by the bftengine code to let implementers of
+the `ICommunication` interface know where received messages should be handled.
+Handlers for received messages implement the `IReceiver` interface. The two sole
+implementers of `IReceiver` are the [ReplicaImp](../bftengine/src/bftengine/ReplicaImp.hpp)
+and the [SimpleClient](../bftengine/src/bftengine/SimpleClient.cpp)
+
+## State transfer
+
+State transfer refers to the mechanism via which out of date replicas can
+retrieve already committed data efficiently and bring themselves in sync with
+the rest of the cluster. State transfer is done outside of the ordering and
+replication protocol performed by SBFT, and can often be achieved by data transfer
+between only two replicas.
+
+As state transfer is outside of the byzantine agreement protocol, it must be
```

Comment:
As state transfer is outside of the byzantine agreement protocol, it must be built as an instantiation function of the specific state machine implementation.